File: rxhttp/src/main/java/com/senierr/http/model/ProgressResponse.java
package com.senierr.http.model;

import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;

/**
 * 带进度的返回
 *
 * @author zhouchunjie
 * @date 2018/8/30
 */
public final class ProgressResponse<T> {

    public static final int TYPE_UPLOAD = -1;
    public static final int TYPE_DOWNLOAD = 1;
    public static final int TYPE_RESULT = 0;

    private final int type;
    private final long totalSize;
    private final long currentSize;
    private final int percent;
    private final @Nullable T result;

    private ProgressResponse(int type, long totalSize, long currentSize, int percent, @Nullable T result) {
        this.type = type;
        this.totalSize = totalSize;
        this.currentSize = currentSize;
        this.percent = percent;
        this.result = result;
    }

    public static @NonNull <T> ProgressResponse<T> upload(long totalSize, long currentSize, int percent) {
        return new ProgressResponse<>(TYPE_UPLOAD, totalSize, currentSize, percent, null);
    }

    public static @NonNull <T> ProgressResponse<T> download(long totalSize, long currentSize, int percent) {
        return new ProgressResponse<>(TYPE_DOWNLOAD, totalSize, currentSize, percent, null);
    }

    public static @NonNull <T> ProgressResponse<T> result(T t) {
        return new ProgressResponse<>(TYPE_RESULT, 0, 0, 0, t);
    }

    public int type() {
        return type;
    }

    public long totalSize() {
        return totalSize;
    }

    public long currentSize() {
        return currentSize;
    }

    public int percent() {
        return percent;
    }

    public @Nullable T result() {
        return result;
    }

    @Override
    public String toString() {
        return "ProgressResponse{" +
                "type=" + type +
                ", totalSize=" + totalSize +
                ", currentSize=" + currentSize +
                ", percent=" + percent +
                ", result=" + result +
                '}';
    }
}

File: rxhttp/src/main/java/com/senierr/http/model/ProgressResponse.kt
package com.senierr.http.model

/**
 * 带进度的返回
 *
 * @author zhouchunjie
 * @date 2018/8/30
 */
class ProgressResponse<T> constructor(
        private val type: Int,
        private val totalSize: Long,
        private val currentSize: Long,
        private val percent: Int,
        private val result: T?
) {
    companion object {

        const val TYPE_UPLOAD = -1
        const val TYPE_DOWNLOAD = 1
        const val TYPE_RESULT = 0

        fun <T> upload(totalSize: Long, currentSize: Long, percent: Int): ProgressResponse<T> {
            return ProgressResponse<T>(TYPE_UPLOAD, totalSize, currentSize, percent, null)
        }

        fun <T> download(totalSize: Long, currentSize: Long, percent: Int): ProgressResponse<T> {
            return ProgressResponse<T>(TYPE_DOWNLOAD, totalSize, currentSize, percent, null)
        }

        fun <T> result(t: T): ProgressResponse<T> {
            return ProgressResponse(TYPE_RESULT, 0, 0, 0, t)
        }
    }

    fun type(): Int {
        return type
    }

    fun totalSize(): Long {
        return totalSize
    }

    fun currentSize(): Long {
        return currentSize
    }

    fun percent(): Int {
        return percent
    }

    fun result(): T? {
        return result
    }

    override fun toString(): String {
        return "ProgressResponse{" +
                "type=" + type +
                ", totalSize=" + totalSize +
                ", currentSize=" + currentSize +
                ", percent=" + percent +
                ", result=" + result +
                '}'.toString()
    }
}